### PR TITLE
[druva-1202] Druva Integration

### PR DESCRIPTION
**Pull request by DruvaPMInstance**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
Druva-Demisto Integration.

## Minimum version of Demisto
- [ ] 4.5.0
- [x ] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x ] No

## Must have
- [x ] Tests
- [x ] Documentation 

## Demisto Partner?
- [x ] The title must be in the following format: **[YOUR_PARTNER_ID] short description**

### CODE REVIEW

**Comment by DruvaPMInstance** [comment]
Note: The below two APIs are currently failing and would be fixed over the coming days.
druva-list-quarantined-snapshots
druva-delete-quarantined-snapshots

File: Packs/Druva/Integrations/Druva/Druva.yml [hunk 1]
```diff
@@ -0,0 +1,209 @@
+commonfields:
+  id: Druva Ransomware Response
+  version: -1
+name: Druva Ransomware Response
+display: Druva Ransomware Response
+category: Data Enrichment & Threat Intelligence
+description: Druva Ransomware Response Playbook
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.yml:7]: update description

File: Packs/Druva/Integrations/Druva/Druva.yml [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  type: 0
+  required: true
+- display: Client ID
+  name: clientId
+  type: 4
+  required: true
+- display: Secret Key
+  name: secretKey
+  type: 4
+  required: true
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.yml:21]: in the description file, add information from where one can retrieve this data

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        raw_response = responseJson
+        return (readable_output, outputs, raw_response)
+    else:
+        demisto.log('Error: ' + str(response.status_code))
+        return None
+
+
+def main():
+    command = demisto.command()
+    params = demisto.params()
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py:265]: avoid calling `demisto.args()` over and over
```suggestion
    params = demisto.params()
    args = demisto.args()
```

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+            search_string = demisto.args().get('search_string')
+            return_outputs(*Druva_FindDevice_Command(clientObj, search_string))
+
+        if command == 'druva-quarantine-resource':
+            resource_id = demisto.args().get('resource_id')
+            resource_type = demisto.args().get('resource_type')
+            from_date = demisto.args().get('from_date')
+            to_date = demisto.args().get('to_date')
+            return_outputs(*Druva_QuarantineResource_Command(clientObj, resource_id, resource_type, from_date, to_date))
+            # return_outputs(*Druva_ListQuarantineRanges_Command(clientObj))
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py:296]: remove all commented code

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+            return_outputs(*Druva_Restore_Status(clientObj, restore_id))
+
+        if command == 'druva-endpoint-decommission':
+            resource_id = demisto.args().get('resource_id')
+            return_outputs(*Druva_Decommission(clientObj, resource_id))
+
+    except Exception as e:
+        return_error('Failed to execute:' + command + 'Error:' + str(e))
+
+    sys.exit(0)
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py:354]: no need

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+    if (statusCode == 200):
+        responseJson = response.json()
+        headers = ['Resource ID']
+        readable_output = tableToMarkdown('Device Decomission Request', str(resource_id), headers=headers)
+        outputs = {"Druva.decomissionedResource(val.resource_id == obj.resource_id)": str(resource_id)}
+        raw_response = responseJson
+        return (readable_output, outputs, raw_response)
+    else:
+        demisto.log('Error: ' + str(response.status_code))
+        return None
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py:260]: better to raise a proper exception with more information on the error (same for all)

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+
+
+def Druva_Decommission(clientObj, resource_id):
+    response = clientObj.post_decommission(resource_id)
+    statusCode = response.status_code
+    responseJson = response.json()
+    if (statusCode == 200):
+        responseJson = response.json()
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py:252]: you have it twice (same for all)
```suggestion
    if (statusCode == 200):
        responseJson = response.json()
```

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+        return "proxy:% s headers:% s" % (self._proxy, self._headers)'''
+
+    def updateHeaders(self, base64String):
+        base64String = base64String.decode("utf-8")
+        headers = {'accept': 'application/json', 'Authorization': 'Basic ' + base64String}
+        data = {'grant_type': 'client_credentials', 'scope': 'read'}
+        response = self._http_request(method='POST', url_suffix='/token', headers=headers, data=data,
+                                      resp_type='response')
+        statusCode = response.status_code
+        if (statusCode == 200):
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py:26]: what happens in case of failure?

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+
+def Druva_DeleteQuarantineRange_Command(clientObj, resource_id, range_id):
+    response = clientObj.delete_quarantineRange(resource_id, range_id)
+    statusCode = response.status_code
+    if (statusCode == 200):
+        responseJson = response.json()
+        headers = ['RangeID']
+        readable_output = tableToMarkdown('Quarantine Range Deleted Successfully', str(range_id), headers=headers)
+        raw_response = responseJson
+        outputs = {"Druva.deletedQuarantineRange(val.rangeID == obj.rangeID)": str(range_id)}
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py:141]: delete should not return context outputs

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+
+
+def Druva_DeleteQuarantined_Snapshots_Command(clientObj, resource_id, range_id, snapshot_id):
+    response = clientObj.delete_Snapshots(resource_id, range_id, snapshot_id)
+    statusCode = response.status_code
+    if (statusCode == 200):
+        responseJson = response.json()
+        headers = ['Snapshot ID']
+        readable_output = tableToMarkdown('Snapshot Deleted successfully', str(snapshot_id), headers=headers)
+        outputs = {"Druva.deletedSnapshots(val.snapshotID == obj.snapshotID)": str(snapshot_id)}
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py:195]: delete should not return context outputs

File: Packs/Druva/pack_metadata.json [hunk 11]
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "Druva",
+    "description": "Druva Integration",
```
[the remark this answers is not in the document]
Comment (sahilgoyaldruva) [reply Packs/Druva/pack_metadata.json]: Requested Changes made.

File: Tests/conf.json [hunk 12]
```diff
@@ -32,6 +32,10 @@
             "playbookID": "FeedMicrosoftIntune_Test",
             "fromversion": "5.5.0"
         },
+{
+    "integrations": "Druva Ransomware Response",
+    "playbookID": "Druva-Test"
+},
```
[the remark this answers is not in the document]
Comment (sahilgoyaldruva) [reply Tests/conf.json]: Requested Changes made.

File: Packs/Druva/Integrations/Druva/README.md [hunk 13]
```diff
@@ -0,0 +1,596 @@
+Druva Ransomware Response Playbook
+This integration was integrated and tested with version xx of Druva Ransomware Response
```
[the remark this answers is not in the document]
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/README.md]: There is no version as it is SaaS product. Changes made to reflect the same.

File: Packs/Druva/pack_metadata.json [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    "support": "partner",
+    "currentVersion": "1.0.0",
+    "author": "Druva",
+    "url": "",
+    "email": "krishnakumar.vasudevan@druva.com",
```
[the remark this answers is not in the document]
Comment (sahilgoyaldruva) [reply Packs/Druva/pack_metadata.json]: Requested Changes made.

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 6 above)
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/Druva.py]: Requested Changes made.

File: Packs/Druva/Integrations/Druva/Druva.yml (hunk 1 above)
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/Druva.yml]: Requested Changes made.

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 7 above)
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/Druva.py]: Requested Changes made.

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 8 above)
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/Druva.py]: Requested Changes made.

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 9 above)
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/Druva.py]: Requested Changes made.

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 10 above)
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/Druva.py]: Requested Changes made.

**Comment by sahilgoyaldruva** [comment]
> nice work
> please update the logo according to standard: https://xsoar.pan.dev/docs/integrations/integration-logo

This has been updated

**Comment by sahilgoyaldruva** [comment]
@yaakovi Changes made. Ready for review

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 15]
```diff
@@ -84,6 +86,15 @@ def post_decommission(self, resource_id):
         return self._http_request(method='POST', url_suffix=url_suffix, resp_type='response')
 
 
+def test_module(clientObj):
+    response = clientObj.test_apiModule()
+    statusCode = response.status_code
+    if (statusCode == 200):
+        return ("ok")
+    else:
+        raise RuntimeError('Error: ' + str(response.status_code))
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py:95]: the error should be more indicative, in this case, words are better than numbers (same for all raised errors).
suggest possible solutions like `Check your credentials` or `Check your connection`...

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 16]
```diff
@@ -293,7 +288,7 @@ def main():
             from_date = demisto.args().get('from_date')
             to_date = demisto.args().get('to_date')
             return_outputs(*Druva_QuarantineResource_Command(clientObj, resource_id, resource_type, from_date, to_date))
-            # return_outputs(*Druva_ListQuarantineRanges_Command(clientObj))
+            return_outputs(*Druva_ListQuarantineRanges_Command(clientObj))
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py]: are you sure about the double `return_outputs`?

File: Packs/Druva/Integrations/Druva/Druva.py [hunk 17]
```diff
@@ -316,7 +311,7 @@ def main():
             return_outputs(*Druva_UpdateQuarantineRange_Command(clientObj,
                                                                 resource_id, resource_type, range_id, from_date,
                                                                 to_date))
-            # return_outputs(*Druva_ListQuarantineRanges_Command(clientObj))
+            return_outputs(*Druva_ListQuarantineRanges_Command(clientObj))
```
Comment (yaakovi) [inline Packs/Druva/Integrations/Druva/Druva.py]: same here

File: Packs/Druva/pack_metadata.json [hunk 18]
```diff
@@ -1,11 +1,11 @@
...
-    "name": "Druva",
-    "description": "Druva Integration",
+    "name": "Druva Ransomware Response",
+    "description": "Druva Ransomware Response Integration provides an API based orchestration framework for Druva Ransomware Recovery customers. The integration allows Druva customers to respond immediately in case of a security incident and recovery their backed up data with Confidence",
     "support": "partner",
     "currentVersion": "1.0.0",
     "author": "Druva",
     "url": "",
-    "email": "krishnakumar.vasudevan@druva.com",
+    "email": "support@druva.com",
```
Comment (yaakovi) [inline Packs/Druva/pack_metadata.json:8]: we added a new option for the partners to monitor changes in their pack.
you can add here a field called `githubUser` with a list value of all Github users you would like to be notified once one of the pack files is changed.

**Comment by yaakovi** [comment]
@sahilgoyaldruva any updates here?

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 16 above)
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/Druva.py]: This was done to return the output as requested in the previous review.

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 17 above)
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/Druva.py]: This was done to return the output as requested in the previous review.

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 15 above)
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/Druva.py:95]: We tried to add that but it wasnt reflecting on the test module. So kept it as is.

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 16 above)
Comment (yaakovi) [reply Packs/Druva/Integrations/Druva/Druva.py]: I meant, if you have the second why do you need the first?

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 17 above)
Comment (yaakovi) [reply Packs/Druva/Integrations/Druva/Druva.py]: I meant, if you have the second why do you need the first?

File: Packs/Druva/Integrations/Druva/Druva.py (hunk 16 above)
Comment (sahilgoyaldruva) [reply Packs/Druva/Integrations/Druva/Druva.py]: the first line indicates success of the operation and second gives more details

**Comment by richardbluestone** [comment]
Reviewed and updated

**Pull request merged**